src/synapse.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python3
+
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Helper module used to manage interactions with Synapse."""
+
+import logging
+import typing
+
+from ops.model import Container
+from ops.pebble import Check, ExecError
+
+from charm_state import CharmState
+from charm_types import ExecResult
+from exceptions import CharmConfigInvalidError, CommandMigrateConfigError
+
+COMMAND_PATH = "/start.py"
+CHECK_READY_NAME = "synapse-ready"
+
+logger = logging.getLogger(__name__)
+
+
+def check_ready(state: CharmState) -> typing.Dict:
+    """Return the Synapse container check.
+
+    Args:
+        state: The state of the charm.
+
+    Returns:
+        Dict: check object converted to its dict representation.
+    """
+    check = Check(CHECK_READY_NAME)
+    check.override = "replace"
+    check.level = "ready"
+    check.tcp = {"port": state.synapse_port}
+    # _CheckDict cannot be imported
+    return check.to_dict()  # type: ignore
+
+
+def synapse_environment(state: CharmState) -> typing.Dict[str, str]:
+    """Generate a environment dictionary from the charm configurations.
+
+    Args:
+        state: The state of the charm.
+
+    Returns:
+        A dictionary representing the Synapse environment variables.
+    """
+    return {
+        "SYNAPSE_SERVER_NAME": f"{state.server_name}",
+        "SYNAPSE_REPORT_STATS": f"{state.report_stats}",
+        "SYNAPSE_NO_TLS": "True",
+    }
+
+
+def is_configuration_valid(state: CharmState) -> bool:
+    """Check if configuration is valid.
+
+    Args:
+        state: The state of the charm.
+
+    Returns:
+        True if they are all set
+    """
+    return all([state.server_name, state.report_stats])
+
+
+def is_server_name_valid(state: CharmState) -> bool:
+    """Check if server name is valid.
+
+    Args:
+        state: The state of the charm.
+
+    Returns:
+        True if server name is valid
+    """
+    return bool(state.server_name)
+
+
+def execute_migrate_config(container: Container, state: CharmState) -> None:
+    """Run the Synapse command migrate_config.
+
+    Args:
+        container: Container of the charm.
+        state: State of the charm.
+
+    Raises:
+        CommandMigrateConfigError: something went wrong running migrate_config.
+        CharmConfigInvalidError: charm configuration is not valid.
+    """
+    if not is_server_name_valid(state):
+        raise CharmConfigInvalidError(
+            "The server_name is empty, please review your charm configuration"
+        )
+    if not is_configuration_valid(state):
+        raise CharmConfigInvalidError(
+            "Configuration is not valid, please review your charm configuration"
+        )
+    migrate_config_command = [COMMAND_PATH, "migrate_config"]
```

Comment:
I would put a TODO comment here about looking into migrating to SDK usage instead of invoking the command

Reply:
I'm not sure if it's possible to use the SDK to generate the configuration file. Maybe using a template?

Reply:
Just in case, I added the comment, thanks!